vowpalwabbit/cache.cc

```diff
@@ -100,68 +157,20 @@ int VW::read_example_from_cache(io_buf& input, example* ae, label_parser& lbl_pa
 
   // read indices
   unsigned char num_indices = input.read_value<unsigned char>("num_indices");
-
-  char* c;
   for (; num_indices > 0; num_indices--)
   {
-    size_t temp;
-    unsigned char index = 0;
-    if ((temp = input.buf_read(c, sizeof(index) + sizeof(size_t))) < sizeof(index) + sizeof(size_t))
-    {
-      VW::io::logger::errlog_error("truncated example! {} {} ", temp, char_size + sizeof(size_t));
-      return 0;
-    }
-
-    index = *reinterpret_cast<unsigned char*>(c);
-    c += sizeof(index);
-
+    unsigned char index;
+    features ours;
+    std::pair<size_t, bool> feat_size_and_sorted = read_cached_index_and_features(input, index, ours);
```

Comment:
this is creating a pair object every time we read an example from cache, we should really check the benchmarks to see if there are any regressions